File: consistent-hashing/consistent-hashing.py
class Solution:
    # @param {int} n a positive integer
    # @return {int[][]} n x 3 matrix
    def consistentHashing(self, n):
        # Write your code here
        segments = []
        segments.append([0, 359, 1])
        for i in range(2, n + 1):
            j = 0
            for index, segment in enumerate(segments[1:]):
                tmp = segment[1] - segment[0]
                len = segments[j][1] - segments[j][0]
                if tmp > len or (tmp == len and segment[2] < segments[j][2]):
                    j = index + 1
            x = segments[j][0]
            y = segments[j][1]
            segments[j][1] = (x + y) / 2
            segments.append([(x + y) / 2 + 1, y, i])
        return segments

File: consistent-hashing/consistent-hashing.cc
class Solution {
public:
    /**
     * @param n a positive integer
     * @return n x 3 matrix
     */

    vector<vector<int>> consistentHashing(int n) {
        // Write your code here
        typedef pair<int, int> pii;
        int total = 360;
        vector<pii> segments(total);
        priority_queue<pii> pq;
        pq.push(make_pair(total, 0));
        segments[0].first = 0;
        segments[0].second = total - 1;
        for (int i = 1; i < n; ++i) {
            int id = -pq.top().second;
            pq.pop();
            segments[i].second = segments[id].second;
            segments[id].second = (segments[id].first+segments[id].second) >> 1;
            segments[i].first = segments[id].second + 1;
            pq.push(make_pair(segments[i].second - segments[i].first + 1, -i));
            pq.push(make_pair(segments[id].second - segments[id].first + 1, -id));
        }
        vector<vector<int>> ans;
        while (!pq.empty()) {
            int id = -pq.top().second;
            pq.pop();
            ans.push_back(vector<int>{segments[id].first, segments[id].second, id + 1});
        }
        return ans;
    }
};

